src/QMCWaveFunctions/WaveFunctionTypes.hpp

```diff
@@ -0,0 +1,49 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers
+//
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//
+// File created by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef QMCPLUSPLUS_QMC_WAVEFUNCTION_TYPES_HPP
+#define QMCPLUSPLUS_QMC_WAVEFUNCTION_TYPES_HPP
+
+#include "type_traits/complex_help.hpp"
+#include "OhmmsPETE/OhmmsMatrix.h"
+
+namespace qmcplusplus
+{
+
+/** Consistent way to get the set of types used in the QMCWaveFunction module, without resorting
+ *  to ifdef'd type aliases accessed through inheritance.
+ *
+ *  You can at least test all flavors without tricky scoping or recompiling.
+ *  Determines type set for class through composition instead of inheritance.
+ *
+ *  Its somewhat question whether it is even worth have all these shortened type aliases
+ *  Defined differently in different modules of the code, but it needs further study.
+ *
+ *  I would have liked to use QMCTypes but they do something dumb with handling complex
+ *  which basically bakes in multiple builds for Real and Complex.
+ */
+template<typename VALUE, typename FP_VALUE>
+struct WaveFunctionTypes final
+{
+  // Should be factored up into a non broken QMCTypes
+  using Value         = VALUE;
+  using FullPrecValue = FP_VALUE;
+  using Real          = RealAlias<Value>;
+  using FullPrecReal  = RealAlias<FullPrecValue>;
+  // This is all that belongs here so far.
```

Comment:
I just feel this is very similar to OrbitalSetTraits.

Reply:
It is but OrbitalSetTraits is quite a bit too complex and I didn't want to fight about it until it could be removed.
I also don't like ScalarTraits

Additionally why does the dirac determinant need to know about the orbitalsettraits which it does via the horrible SPOSt::type stuff?  I just needs the right value type to be returned by the function calls to Phi.  SPOSet should pretty obviously be templated on Value but I don't think that's in scope for this PR.

Reply:
I think we have more understanding on this. 
WaveFunctionTypes is intended for making WFC derived classes, at least right now, more towards template.